### ISSUE TITLE
Security To-do

### ISSUE THREAD

**Issue by Apogate** [spootTheLousy/saguaro#138]
Saguaro has some concerning security issues that need to be addressed, mostly revolving around **admin.php**.

Current staff-user authentication relies on 2 cookies, username and user-specific password. The password cookie is plain-text. Really bad. These cookies are not sent over ssl and are vulnerable to theft. There is no current limit on logins, and the panel form is vulnerable to (somewhat determined) bruteforcing attacks. Passwords are stored in the database in plain text. **BAD.**

**Proposal:** System that uses 3 hashed cookies for authentication, one being constant, two being user name/password values to verify actions.

`PANEL_PASS` would be the server "key", staff username/password would be client end keys, this would also tie in nicely with the new user class by using session cookies.

**Saltfile**
Last I checked, Saguaro generates a salt file upon the first use of a secure tripcode, I propose changing this to be generated by test.php using a user-supplied string in the config. This would be used by:
- **Tripcode.php** (Secure tripcodes)
- **User.php** (Hashing passwords) 
- **Admin.php** (Verifying logins)

<h4>Checklist:</h4>

<hr>

**Password salting**
- [x] Hash passwords stored in database
- [x] Hash cookie values
  - **PARTIAL** Indirect result of hashing stored passwords.
    <hr>

**Admin.php**
- [ ] Add cookie `saguaro_janitoken`, hashed with saltfile
- [ ] Set cookie flags `SECURE` and `HTTPONLY` for staff cookies `saguaro_auser` and `saguaro_apass`.
  - This would only allow cookies to be set via https requests, I don't know what the implications are for users whos host doesn't have ssl @RePod pls enlighten
- [x] Limit login form attempts
  - ~~**Not to be done done via captcha, I'll figure something else out for this.**~~
  - 6a68de7f5b6f4ba1ea3ec320235d81950fb17825 "Did it via captcha anyway." - RePod
    <hr>

**User.php**
- [x] Add user creation function (used by admin.php)
- [x] Add user deletion function (used by admin.php)
  <hr>

**Config.php**
- [ ] `SALTFILE_PATH` added, filepath to salt generated
- [ ] `SALT` added, this would be a string used to generate the saltfile
  <hr>

**Test.php**
- [ ] Check if `SALTFILE_PATH` and `SALT` have been modified from default values
- [ ] Generate salt file
  <hr>

**Tripcode.php**
- [x] Re-reference new salt file for use in generating secure tripcodes 33afd0682f651e5c140d0422240d07643867b35b
  - This removes the use of the leaked yotsuba-"LOLOLOLOLOLO" salt used.
    <hr>

**Comment by RePod** [comment]
## User.php

I don't think these commands belong here, other than that they're "relevant". The User class is intended to be a single point of reference to a User's status and what they can do (not will do, like post).

While it could be modified to handle user **_creation**_ (accounts) and additional processing, that wasn't the original intention.

Anything that touches the SQL database (specifically writing) should be considered **administrative commands** for all intents and purposes and be handled separately.

## General Security
- Management does not need to be accessible from the board. I'm not saying security through obscurity but if they can't find the login page (especially when they certainly can't login) that's the first step.
  - May annoy casual users when they don't know where to find it, but this would definitely be better than what it is right now.
  - Should also be unique to each server, and possibly board. Not `site.com/board/admin` or similar.
- Rate limiting would be difficult/impossible if using sessions/cookies which means we'd have to it with the SQL database. Fortunately there's already a _loginattempts_ table to work with.
  - Implement CAPTCHA. **DONE** 6a68de7f5b6f4ba1ea3ec320235d81950fb17825
  - However this could also lead to the problem of constantly hitting the SQL server so a solution on the PHP server would also need to be done to completely lock them out.
  - Perhaps salt a unique identifier about the user request (like IP) and store it somewhere on the PHP server, check if it exists then die.
- Finish converting residue files to classes to prevent arbitrary execution from direct access.

## Password Salting (done: 86689869bdef93f2974f601490a488973868f06c)

I am all for this, but there are a few problems with this concept:
- Need to keep parity between the client sending the info. **OR NOT!**
  - Needs to be hashed client side before sending if we don't want to send plain-text.
  - Reveals the hashing algorithm. (requires client side scripting)
  - Unless a basic hash (like just to md5) is done and then it is processed further on the server, skipping a step but making it harder to keep parity.
  - In environments that actually want to secure their login, they're almost always over HTTPS.
  - See third point for first resource below.
- Non-reversible.
  - Most people would be fine with this but extreme casuals might not like how hard it is to change their password without re-hashing it or opening _test.php_.

For the back-end of things, using [cryptically (whatever) secure functions](https://crackstation.net/hashing-security.htm) intended for hashing along with proper salting will be good.

Fortunately, since we're only generating the passwords once (currently) in _test.php_ we can be as slow possible to ensure they're thoroughly secured (see first resource on "Key Stretching"). Also with verification to re-hash the given password, since we're not on a huge scale and the number of actual logins should be considerably less than actual users posting we can be slow here too.

**NOTE:** It is pointed out (in my first resource) that while slower is better, the obvious CPU utilization would quickly add up across multiple legit uses and exponentially while attacking resulting in effectively a DDoS (overloading the CPU). We can mitigate a majority of the problem with rate limiting but it's something to consider for hashing speed.

## Authenticating

Users should provide their login name and password as normal, but only store a separate hash returned by the server (stored in the SQL database) to validate them later. As for expiring them automatically, that's another story. Logging out would be one way.

For processing, all that would need to be checked is the hash in the cookie. Security-wise it's not more secure than storing anything else if that's all there is to it, but it's better than storing the username/password.

## Configurable salt

Nothing really to say to this that isn't covered by my resource(s).

The gist is everything should be uniquely salted so duplicated content (like two "hello" passwords) don't share the same hash.

To verify it would mean storing the salt it was salted with, so it would be like: `CONFIG_SALT + UNIQUE_SALT + data_to_salt`.  UNIQUE_SALT does not need to be kept secret. However, based on my reading a configurable salt at all is kind of useless since it's effectively hard-coded and in extreme cases if it were to leak it'd be useless (but this is still an option).

The issue with the configurable salt at this point would be since we're using it with another unique salt (that's also being stored but alongside its output) it's _almost_ pointless if the user does not change it. _test.php_ would need to be updated to actually request (basically a minor rewrite to make it a full installation script) and guarantee there's a salt.

**TL;DR:** Using both a secret, configurable salt and unique salt for hashing as one "big" salt is fine.
The unique salt is way more important though.

It'd basically be double-salting, which shouldn't be a huge problem. 
- Result of `CONFIG_SALT` and `UNIQUE_SALT` becomes the salt for for `data_to_salt`
- Previous example, `CONFIG_SALT + UNIQUE_SALT + data_to_salt`.

## Final notes

All reading/writing/hashing/verification should be done from a single point and not be spread among multiple files (other than class extending). Everything needing to utilize these features should work alongside the specific point in question and let it handle it instead of implementing itself. This will lead to consistency across the board and a single point of focus.

## Reading material / Resources

[Secure Salted Password Hashing - How to do it Properly](https://crackstation.net/hashing-security.htm)
- Proper salting and shortcomings of bad salting.
- Handling client-side hashing (basically never), and should be done server-side.
- Insecurities of non-HTTPS logins (specifically client-side). While we still _can_ use HTTP, we should place good faith in the golden rule of never trusting the client. It'd make no difference if we send it pre-hashed or plain-text over HTTP since they can both be intercepted, modified, and abused.
- Dictionary/brute force attacks cannot be completely stopped, only hindered.
- Use vague wording to obscure the existence of a valid username, "Invalid username **_or**_ password." instead of "Invalid username.". **PARTIAL** 
  - aa494fa0ce4fdbfe80fd2a2370cc4cba671dc4b0 Vague wording.
  - c807ef179896e42bb7176c339e237c70e0979680 Hide generic messages.

**Comment by RePod** [comment]
> "I exist to upset Apogate by going against everything he says."
> -- RePod, 2015.

6a68de7f5b6f4ba1ea3ec320235d81950fb17825 implements stock captcha or reCAPTCHA (if enabled) on the management login screen.
Redesigned the form so it wasn't an eyesore.

**The captcha's presence (not reCAPTCHA) cannot be disabled via config.**
However, it's relatively simple to comment it out in Login. If you want to make it toggle-able from config I'd recommend enabling it by default, _SECURE_LOGIN_ or something.

![screenshot_1](https://cloud.githubusercontent.com/assets/7897871/10843987/c89171e2-7ed2-11e5-8cc3-43f970edd7da.jpg)

Also Login->auth() is a mess. It's doing a lot more than generating the login page, that should be changed @Apogate.

My thoughts:
- This could be modified to only occur after N failed attempts.
  - If we count via session it can easily be avoided by clearing cookies.
  - If we count server-side we're infinitely increasing resource usage.
- Either captcha (more so reCAPTCHA) increases the odds of it not being automated.
- Captchas shouldn't be a problem with valid users/logins as they should get it first try.
  - Especially with reCAPTCHA since in the best scenario only one click is required.
  - This only needs to be done when logging in, which is persistent afterwards.
  - Basic rate limiter and requires to be solved (cannot be forged/bypassed, even on stock).
  - Stock can easily be automated (unless improved), which is why reCAPTCHA is preferred.
- Neither captcha requires interfacing with the SQL server (and halts execution if failed).

**Comment by RePod** [comment]
SaguaroCrypt/Legacy is available for testing: ab5e3ca877df7d69400e3d897cc70dd9698fffb3

Tests are included in the head comment, notably:
- Generating the content to store in the SQL database (hash and public salt).
- Comparing a given input with an existing hash and public salt (read from SQL db).

Relocating the tests to after the class and accessing the same file is good enough.

**A configurable salt is currently not implemented and used in the public salt generation or hashing processes.**

This is intended for _legacy_ PHP (pre-5.3), but not tested for it.
5.3+ and 7 specifically added additional functions and security fixes that will eventually be made into SaguaroCrypt/Modern.

**Not audited.**

## Implementing

**_test.php**_ will need to load _SaguaroCrypt_ and use it to hash the passwords to be written to the SQL database:

``` PHP
SaguaroCryptLegacy->generateHash('muhpassword');
```

This will return the resulting hash and public salt used.

**_admin/login.php**_ will need to load _SaguaroCrypt_ and use it to verify the given password with the retrieved hash and public salt:

``` PHP
SaguaroCryptLegacy->compare_hash('givenpassword', 'storedhash', 'storedpublicsalt');
```

This will return either `true` or `false`.

The user table may need to be modified and break legacy support.
Adding a new salt column would be best:

| username | password | salt | allowed | denied |
| --- | --- | --- | --- | --- |

**_test.php**_ _can_ be modified to update old tables.

Alternatively, users can be requested to drop their old logins table when upgrading to be able to login. This would be easier.

_Another_ alternative is:
- **_test.php**_ caching the entire logins table (partially insecure).
  - Updating legacy rows as needed.
- Dropping the old table.
- Writing out the updated cache to a new table.
- Every time it is run.

**Comment by RePod** [comment]
86689869bdef93f2974f601490a488973868f06c Relevant files have been updated to take advantage of SaguaroCrypt/Legacy, specifically **_test.php**_ and **_admin/login.php**_. For specifics see the commit message.

@Yushe will be pleased to know passwords are now encrypted in the database. However, the logins table needs to be dropped. Debating adding legacy upgrade support, definitely need more opinions on this.

There is a minor performance hit when logging in due to the verification, but it should be negligible for low usage.

This indirectly _almost_ solves the issue of **storing** the password as plain-text in the cookies, except now it's storing the hashed password (which is useless without the salt). Still an improvement.

As for alleviating this, I don't know why the username and (now hashed) password are being stored at all. I brought this up above in my "Authentication" section where the server should assign a unique hash. The chances of duplicating it are slim, and again would make no difference in snooping situations.

**Comment by RePod** [comment]
**Cookies:**
- `Secure` should not be used.
- `HttpOnly` definitely should be used.
  - Added in PHP 5.2.0

I have no formal logic for this, but if our current concern is catering to both HTTP and HTTPS `Secure` will do absolutely nothing. `HttpOnly` is a generic security improvement as it restricts what can access the cookie to actual requests (i.e.: not with Javascript and `document.cookie`).

Since it's restricted to admin logins, if the site provides an SSL connection they should be prioritizing it over non-SSL.

**Comment by Apogate** [comment]
Notes for me:
login issues:

- [ ] how do i set public salt lol
- [ ] loginattempts primary key errors

**Comment by RePod** [comment]
1b06552df7e449f91c26772127295c3764f6aa1c is a disaster.

Commented problems and solutions.

**Comment by Apogate** [comment]
maybe we shouldn't be writing the salt to a file. maybe that's a bad idea

**Comment by Apogate** [comment]
Muh [CSRF](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Introduction)
- [ ] Tweak `valid()`
- [ ] Either implement or axe `_core/user/user.php`

**Comment by Apogate** [comment]
[We may still have potential for SQL injection apparently.](http://stackoverflow.com/questions/60174/how-can-i-prevent-sql-injection-in-php/8255054#8255054)
we're kusaba tier until changes are implemented to SQL once and 4 all. it's that time of year again, SQL rewrites are on the horizon once more.

prepared statements ahoy

**Comment by RePod** [comment]
Daily reminder the login page is still easily accessible from the board.

Granted it wouldn't take much to find it otherwise.

**Comment by RePod** [comment]
Other SQL security issues mostly revolve around sanitation.

I haven't looked at the sanitize class much, but obvious things like enforcing only alpha-numeric characters, etc. (via regex, `/\w+/i`) and reasonable length limits (to prevent long arbitrary strings in places other than the comment) would help.

The downside to length limits is that it removes breathing room for legit users.

Assume somebody was using a long name `Anonymously Anonymous` (21 characters) and a secure tripcode (2 (##)+? characters). If the length limit for the name field was enforced at 30 characters before being processed, the maximum length of the actual tripcode could be no more than 7 characters (which itself is decent entropy as a trade off for such a long name + secure salt). A name field of `##supersecrettripcodeguessme` (2+26 characters) would give good entropy at the trade off of a 0-2 character actual name.

Via regex, the check would be similar to `/[#\w ]{,30}/i`.

> Why not treat names and tripcodes seperately (before processing)?

Possible, and probably not difficult.
We'd have to check the supplied name earlier and verify each section individually.
After that, we need to make sure we can store it all. Currently the `name` column has no "defined" limit other than its column type `text` (_up to_ 65k characters) and also stores more than just the name, such as formatted trip/capcodes. Because we're arbitrarily storing more than just the name (such as HTML formatting+*codes+etc), it's hard to establish an upper limit to this column. However, by adding such a limit it makes it harder to store arbitrary code in places other than the `comment` column.

As it stands, there is currently no limit to:
- Name field (client or server)
  - `text` column type, up to 65k characters.
- E-mail field (client or server)
  - `text` column type, up to 65k characters.
- Password field/`pwd` column (server).
  - Forging a longer password than allowed (8 characters) results in an empty value for the row's `pwd` column.
  - `text` column type, up to 65k characters.
- Comment/`com` column (client OR server)
  - `text` column type, up to 65k characters. **_[Actually possible.](https://cloud.githubusercontent.com/assets/7897871/13679807/a2be5e2c-e6c3-11e5-8c67-d026814bf4fc.jpg)**_

These may be byproducts of the new Regist, however limiting the columns themselves would help.
Excess of a column type's allowed length seems to be gracefully trimmed. However, we shouldn't abuse this behavior to avoid trimming it ourselves before using it in a query.

**Comment by RePod** [comment]
For uploaded files, which may contain harmful contents other than intended we have a few options:
- `getimagesize(file)` which returns the dimensions.
  - We currently already do this, so it's here to show its hierarchy status.
  - Limited to images. WebMs use _avprobe_ to determine resolution.
  - Couple this with the pre-existing config defines `MIN_W` and `MIN_H`
  - Obviously a 0x0 file isn't particularly what we want either.
- Check the MIME type correctly using [finfo_file](//php.net/manual/en/function.finfo-file.php).
  - This might need to be done first to load the correct processor instead of guessing based on the extension.
  - Two tiers to this, not an image at all and not the proper MIME+extension.
- [Attempt to create a usable image from it from scratch.](//php.net/manual/en/function.finfo-file.php#75275)
  - Most computationally (CPU+Memory) intense.
  - Pretty much the catch all.

For images specifically, I think the best bet would be to impose these checks in increasing severity if they were passed. For instance, if `getimagesize()` returns bad for either axis stop immediately otherwise proceed to the MIME check, then finally recreating it.

**Comment by Apogate** [comment]
saltfile needs more pathing options because i just downloaded all of yushes salts and am currently decrypting his database to triangulate his position in a visual basic gui

**Comment by RePod** [comment]
Switching to *at least* MySQLi and prepared statements ***WHEN?***

**Comment by RePod** [comment]
This was for more than SQL but okay.
